generator/internal/mixin_utils.cc

```diff
@@ -49,61 +49,80 @@ std::unordered_map<std::string, std::string> const& GetMixinProtoPathMap() {
+    } else if (rule_key == "post") {
+      http_rule.set_post(rule_value);
+    } else if (rule_key == "put") {
+      http_rule.set_put(rule_value);
+    } else if (rule_key == "patch") {
+      http_rule.set_patch(rule_value);
+    } else if (rule_key == "delete") {
+      http_rule.set_delete_(rule_value);
+    }
+  }
 }
 
 /**
  * Extract Mixin methods from the YAML file, together with the overwritten http
  * info.
  */
-std::unordered_map<std::string, MixinMethodOverride> GetMixinMethodOverrides(
+std::unordered_map<std::string, google::api::HttpRule> GetMixinHttpOverrides(
     YAML::Node const& service_config) {
-  std::unordered_map<std::string, MixinMethodOverride> mixin_method_overrides;
-  if (service_config.Type() != YAML::NodeType::Map)
-    return mixin_method_overrides;
-  if (!service_config["http"]) return mixin_method_overrides;
+  std::unordered_map<std::string, google::api::HttpRule> http_rules;
+
+  if (service_config.Type() != YAML::NodeType::Map) return http_rules;
+  if (!service_config["http"]) return http_rules;
+
   auto const& http = service_config["http"];
-  if (http.Type() != YAML::NodeType::Map) return mixin_method_overrides;
+  if (http.Type() != YAML::NodeType::Map) return http_rules;
+
   auto const& rules = http["rules"];
-  if (rules.Type() != YAML::NodeType::Sequence) return mixin_method_overrides;
+  if (rules.Type() != YAML::NodeType::Sequence) return http_rules;
+
   for (auto const& rule : rules) {
     if (rule.Type() != YAML::NodeType::Map) continue;
 
     auto const& selector = rule["selector"];
     if (selector.Type() != YAML::NodeType::Scalar) continue;
-    std::string const method_full_name = selector.as<std::string>();
 
-    for (auto const& kv : rule) {
-      if (kv.first.Type() != YAML::NodeType::Scalar ||
-          kv.second.Type() != YAML::NodeType::Scalar)
-        continue;
-
-      std::string const http_verb_lower =
-          absl::AsciiStrToLower(kv.first.as<std::string>());
-      auto const& http_verbs = GetHttpVerbs();
-      auto const it = http_verbs.find(http_verb_lower);
-      if (it == http_verbs.end()) continue;
-      std::string const& http_verb = it->second;
-      std::string const http_path = kv.second.as<std::string>();
-      absl::optional<std::string> http_body;
-      if (rule["body"]) {
-        http_body = rule["body"].as<std::string>();
+    std::string const method_full_name = selector.as<std::string>();
+    google::api::HttpRule http_rule;
+    ParseHttpRule(method_full_name, rule, http_rule);
+
+    if (rule["additional_bindings"]) {
+      auto const& additional_bindings = rule["additional_bindings"];
+      if (additional_bindings.Type() == YAML::NodeType::Sequence) {
+        for (auto const& additional_binding : additional_bindings) {
+          google::api::HttpRule& additional_http_rule =
+              *http_rule.add_additional_bindings();
+          if (additional_binding.Type() != YAML::NodeType::Map) continue;
```

Comment:
I don't think we will ever encounter malformed YAML, but this `continue` would have to happen before we add an `additional_binding` to `http_rule`.

Reply:
Good catch!